python/ray/util/iter.py

```diff
@@ -279,6 +280,81 @@ def local_shuffle(self, shuffle_buffer_size: int,
                 str(seed) if seed is not None else "None"))
 
+    def repartition(self, num_partitions: int) -> "ParallelIterator[T]":
+        """Returns a new ParallelIterator instance with num_partitions shards.
+
+        The new iterator contains the same data in this instance except with
+        num_partitions shards. The data is split in round-robin fashion for
+        the new ParallelIterator.
+
+        Args:
+            num_partitions (int): The number of shards to use for the new
+                ParallelIterator
+
+        Returns:
+            A ParallelIterator with num_partitions number of shards and the
+            data of this ParallelIterator split round-robin among the new
+            number of shards.
+
+        Examples:
+            >>> it = from_range(8, 2)
+            >>> it = it.repartition(3)
+            >>> list(it.get_shard(0))
+            [0, 4, 3, 7]
+            >>> list(it.get_shard(1))
+            [1, 5]
+            >>> list(it.get_shard(2))
+            [2, 6]
+        """
+
+        def base_iterator(all_actors,
+                          num_partitions,
+                          partition_index,
+                          timeout=None):
+            futures = {}
+            for a in all_actors:
+                futures[a.par_iter_slice.remote(
+                    step=num_partitions, start=partition_index)] = a
+            while futures:
+                pending = list(futures)
+                if timeout is None:
+                    # First try to do a batch wait for efficiency.
+                    ready, _ = ray.wait(
+                        pending, num_returns=len(pending), timeout=0)
+                    # Fall back to a blocking wait.
+                    if not ready:
+                        ready, _ = ray.wait(pending, num_returns=1)
+                else:
+                    ready, _ = ray.wait(
+                        pending, num_returns=len(pending), timeout=timeout)
+                for obj_id in ready:
+                    actor = futures.pop(obj_id)
+                    try:
+                        yield ray.get(obj_id)
+                        futures[actor.par_iter_slice.remote(
+                            step=num_partitions,
+                            start=partition_index)] = actor
+                    except StopIteration:
+                        pass
+                # Always yield after each round of wait with timeout.
+                if timeout is not None:
+                    yield _NextValueNotReady()
+
+        # initialize the local iterators for all the actors
+        all_actors = []
+        for actor_set in self.actor_sets:
+            actor_set.init_actors()
+            all_actors.extend(actor_set.actors)
+
+        def make_gen_i(i):
+            return lambda: base_iterator(all_actors, num_partitions, i)
+
+        generators = [make_gen_i(s) for s in range(num_partitions)]
+        return from_iterators(
+            generators,
```

Comment:
Should this keep a reference to the original iterator / actors? Seems like it is only indirectly referenced through the generator object now.